### PR TITLE
microservices: add tenant spec

### PR DESCRIPTION
**Pull request by frzifus**

### CODE REVIEW

File: apis/tempo/v1alpha1/microservices_types.go [hunk 1]
```diff
@@ -86,6 +93,188 @@ type MicroservicesStatus struct {
...
+// RoleSpec describes a set of permissions to interact with a tenant.
+type RoleSpec struct {
+	Name        string           `json:"name"`
+	Resources   []string         `json:"resources"`
+	Tenants     []string         `json:"tenants"`
+	Permissions []PermissionType `json:"permissions"`
+}
+
+// OPASpec defines the opa configuration spec for Tempo Gateway component.
+type OPASpec struct {
```
Comment (pavolloffay) [inline apis/tempo/v1alpha1/microservices_types.go:204]: can be removed for now

File: apis/tempo/v1alpha1/microservices_types.go [hunk 2]
```diff
@@ -86,6 +93,188 @@ type MicroservicesStatus struct {
...
+	//
+	// +required
+	// +kubebuilder:validation:Required
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="OpenPolicyAgent URL"
+	URL string `json:"url"`
+}
+
+// TenantSecretSpec is a secret reference containing name only
+// for a secret living in the same namespace as the (Tempo) Microservices custom resource.
+type TenantSecretSpec struct {
```
Comment (pavolloffay) [inline apis/tempo/v1alpha1/microservices_types.go:215]: can be removed for now

Comment (pavolloffay) [reply apis/tempo/v1alpha1/microservices_types.go:215]: What is it used for?

**Comment by pavolloffay** [comment]
CI failed

**Review by iblancasa — CHANGES_REQUESTED** [review]

File: apis/tempo/v1alpha1/microservices_types.go (hunk 2 above)
Comment (frzifus) [reply apis/tempo/v1alpha1/microservices_types.go:215]: `Secret defines the spec for the clientID, clientSecret and issuerCAPath for tenant's authentication` its there to folllow the structure of the loki-operator. But we could also replace it with a [v1.Secret](https://kubernetes.io/docs/concepts/configuration/secret/).

File: apis/tempo/v1alpha1/microservices_types.go [hunk 3]
```diff
@@ -86,6 +93,188 @@ type MicroservicesStatus struct {
...
+// ModeType is the authentication/authorization mode in which Tempo Gateway
+// will be configured.
+//
+// +kubebuilder:validation:Enum=openshift
+type ModeType string
+
+const (
+	// Dynamic mode delegates the authorization to a third-party OPA-compatible endpoint.
+	Dynamic ModeType = "dynamic"
+	// Static mode asserts the Authorization Spec's Roles and RoleBindings
```
[the remark this answers is not in the document]
Comment (frzifus) [reply apis/tempo/v1alpha1/microservices_types.go]: the comment was a mistake.

File: apis/tempo/v1alpha1/microservices_types.go [hunk 4]
```diff
@@ -76,6 +76,63 @@ type MicroservicesSpec struct {
...
+
+// ModeType is the authentication/authorization mode in which Tempo Gateway
+// will be configured.
+//
+// +kubebuilder:validation:Enum=dynamic
+type ModeType string
+
+const (
+	// Dynamic mode delegates the authorization to a third-party OPA-compatible endpoint.
+	Dynamic ModeType = "dynamic"
```
Comment (iblancasa) [inline apis/tempo/v1alpha1/microservices_types.go:110]: Are there any other `ModeType` values? Do we think to support them?

**Review by iblancasa — CHANGES_REQUESTED** [review]
Since (from the name of the branch) this PR is to support authorization in the CRD. Is this feature "complete" (I can enable authentication in a Tempo instance)? If yes, would it make sense to add an E2E test?

File: apis/tempo/v1alpha1/microservices_types.go (hunk 4 above)
Comment (frzifus) [reply apis/tempo/v1alpha1/microservices_types.go:110]: yes, there are more coming.

**Comment by frzifus** [comment]
> Since (from the name of the branch) this PR is to support authorization in the CRD. Is this feature "complete" (I can enable authentication in a Tempo instance)? If yes, would it make sense to add an E2E test?

Nope, it just extends the CRD to avoid conflicts between pavols and my work on that end.

**Review by pavolloffay — APPROVED** [review]

**Comment by pavolloffay** [comment]
Merging this to avoid conflicts and progress with the work. Any changes can be made in the follow up PRs.

**Comment by pavolloffay** [comment]
> Since (from the name of the branch) this PR is to support authorization in the CRD. Is this feature "complete" (I can enable authentication in a Tempo instance)? If yes, would it make sense to add an E2E test?


e2e tests will be added in a follow up PRs. This feature will be quite big.

**Pull request merged**